@here/harp-mapview/lib/Extrusion.ts

```diff
@@ -0,0 +1,182 @@
+
+    static DEFAULT_EXTRUSION_HEIGHT_MIN: number = 0.01;
+    static DEFAULT_EXTRUSION_HEIGHT_MAX: number = 1;
+
+    private m_zoomLevel: number;
+    private m_zoomLevelPrevious: number;
+    private m_tilesSet: Map<Tile, AnimationParams> = new Map();
+    private m_zoomedIn: boolean = false;
+
+    constructor(private mapView: MapView) {
+        this.m_zoomLevelPrevious = mapView.zoomLevel;
+        this.m_zoomLevel = mapView.zoomLevel;
+    }
+
+    add(
+        tile: Tile,
+        options: {
+            animatonEnabled: boolean;
+            animatonDuration: number;
+            animatonZoomLevel: number;
+        }
+    ): void {
+        this.m_tilesSet.set(tile, {
+            animationState: AnimationState.None,
+            animationDuration: options.animatonDuration,
+            extrusionHeight: Extrusion.DEFAULT_EXTRUSION_HEIGHT_MIN
+        });
+
+        tile.extrusionHeight = Extrusion.DEFAULT_EXTRUSION_HEIGHT_MIN;
+        this.startExtrusionAnimation();
+    }
+
+    remove(tile: Tile): void {
+        this.stopExtrusionAnimation();
+        this.m_tilesSet.delete(tile);
+    }
+
+    set zoom(zoomLevel: number) {
+        this.m_zoomLevelPrevious = this.m_zoomLevel;
+        this.m_zoomLevel = zoomLevel;
+
+        if (this.m_zoomLevelPrevious !== this.m_zoomLevel) {
+            if (
+                this.m_zoomLevel >= Extrusion.DEFAULT_EXTRUSION_ANIMATION_ZOOM_LEVEL &&
+                this.m_zoomedIn === false
+            ) {
+                this.m_zoomedIn = true;
+                this.startExtrusionAnimation();
+            }
+
+            if (
+                this.m_zoomLevel < Extrusion.DEFAULT_EXTRUSION_ANIMATION_ZOOM_LEVEL &&
+                this.m_zoomedIn === true
+            ) {
+                this.m_zoomedIn = false;
+                this.m_tilesSet.forEach(animationParams => {
+                    animationParams.extrusionHeight = Extrusion.DEFAULT_EXTRUSION_HEIGHT_MIN;
+                    animationParams.animationState = AnimationState.None;
+                    this.updateHeightForLoadedTiles();
+                });
+
+                this.stopExtrusionAnimation();
+            }
+        }
+    }
+
+    private startExtrusionAnimation(forceAnimate?: boolean): void {
+        this.stopExtrusionAnimation();
+
+        this.m_tilesSet.forEach(animationParams => {
+            if (animationParams.animationState === AnimationState.None) {
+                animationParams.animationState = AnimationState.Started;
+            }
+        });
+
+        this.animate();
+
+        this.updateHeightForLoadedTiles();
+
+        this.mapView.addEventListener(MapViewEventNames.AfterRender, this.animate);
+    }
+
+    private stopExtrusionAnimation(): void {
+        this.mapView.removeEventListener(MapViewEventNames.AfterRender, this.animate);
+
+        this.m_tilesSet.forEach(
+            animationParams => (animationParams.animationStartTime = undefined)
+        );
+    }
+
```

Comment:
`startExtrusionAnimation` and `stopExtrusionAnimation` are then common to all tiles? I can see that all buildings are going up at the same time. When a new tile is loaded in between, `stopExtrusionAnimation` is called, which brings the building down, etc. until all tiles are loaded, and finally the buildings can reach their correct height.

It seems instead that these function should be independent between tiles.

In some previous reviews there were worries that adding up `requestAnimationFrame` for each tile was a problem from the framerate, but in fact it is just a callback that the browser calls when a frame is ready. When you add the unique listener to `AfterRender`, it is the same, and we could have as many as these small callbacks as there are tiles in the view. It should not be a performance problem.